Add system for Dynamic Registry Flags

rinOfTheStars opened the pull request:
This PR adds the beginnings of a system similar to (but far simpler than) the Static Registry Sync API for Dynamic Registries, allowing flags to be set which modify the behavior of Dynamic Registries.

Currently, only one flag, `DynamicRegistryFlag.OPTIONAL`, exists—this flag is a clone of the similar `SyncOption#skipWhenEmpty`, and thus only effects synchronized dynamic registries, and simply has no effect on non-synced dyn. registries. 

Both the `OPTIONAL` flag (and it's FAPI counterpart) set whether a dynamic registry should attempt to be synced with a connecting client which does not have said dynamic registry. This allows for clients without certain dynamic registries to still be able to connect, and thus is useful when creating entirely server-sided mods intended to be compatible with vanilla.

More flags can easily be added in the future, however the infrastructure for flags that apply to *all* dynamic registries, and not just synced ones, has not yet been set up—a future PR to add such a flag would need to add a mixin for that purpose somewhere other than `DynamicRegistrySync`. 

Note that the test mod for the registry API is ironically too complicated to serve as a proper test for this API addition—I have created a test mod locally (using the `mavenLocal()` repo) and tested having a vanilla client connect to a pure QSL + testmod server with an optional dynamic registry defined in the testmod, and it thankfully did work. If required, I can upload this testmod's repo to github with instructions on how to run it.

rinOfTheStars committed:
Implement Dynamic Registry Flags.
- Currently, only one flag, `OPTIONAL`, exists.
- Modified DynamicRegistrySyncMixin in order to implement these flags. Due to the Redirect mixin header used to implement the flag filtering being taken directly from FAPI, the header of DynamicRegistrySyncMixin was changed to mention FabricMC copyright, in accordance with QSL contribution rules in CONTRIBUTING.md

rinOfTheStars committed:
Implement Dynamic Registry Flags p2
- Modify DynamicRegistryFlagManager to use Identifiers over RegistryKeys internally, in order to avoid generic-related jank.
- Fix mixin ported from FAPI by correcting method target string to use Quilt Mappings.
- Implement very basic tests.

rinOfTheStars committed:
Expanded upon OPTIONAL flag javadocs to elaborate use-case

rinOfTheStars committed:
Expanded further on OPTIONAL javadocs

rinOfTheStars commented:
This should probably be considered a blocking addition for FAPI 0.86.0 & 0.86.1 compat in QFAPI alongside #329.

Pyrofab commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/dynamic/DynamicRegistryFlag.java:
```diff
@@ -0,0 +1,82 @@
+ */
+public enum DynamicRegistryFlag {
+	/**
+	 * Note: This flag is intended only for synchronized dynamic registries. On non-synced dynamic registries, this flag does nothing.
```
I believe this note should come second - the short description should always be on the first line

Pyrofab commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/dynamic/DynamicRegistryFlag.java:
```diff
@@ -0,0 +1,82 @@
+	 * Indicates that this registry (and the entries within) do not necessarily need to be sent
+	 * to (logical) clients for synchronization in multiplayer contexts.
+	 * <p>
+	 * This is useful, for instance, when creating a server-sided mod that is vanilla client compatible when installed on a server,
+	 * but has extra functionality for players when the mod is installed on a connecting client.
```
Could possibly be an `@apiNote`, not sure if we use that in QSL

rinOfTheStars replied:
Doesn't seem to exist in QSL.

Pyrofab commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/dynamic/DynamicRegistryFlag.java:
```diff
@@ -0,0 +1,82 @@
+	 * @param registryId the value id ({@link RegistryKey#getValue()}) of the target dynamic registry
+	 * @param flag the flag value to enable on the dynamic registry
+	 */
+	public static void enableFlag(Identifier registryId, DynamicRegistryFlag flag) {
```
Wouldn't it make more sense / be more discoverable to have the flags be passed either as a vararg or as an `Set` in `DynamicMetaRegistry.registerSynced` ?

Pyrofab commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/dynamic/DynamicRegistryFlag.java:
```diff
@@ -0,0 +1,82 @@
+	 * @param registryId the value id ({@link RegistryKey#getValue()}) of the target dynamic registry
+	 * @param flag the flag value to disable on the dynamic registry
+	 */
+	public static void disableFlag(Identifier registryId, DynamicRegistryFlag flag) {
```
In which case would the server disable a flag ? Furthermore, isn't there a risk of causing desyncs if a flag is disabled after the server finishes loading ?

rinOfTheStars replied:
Resolved in latest commit. Decided to remove the ability to disable flags, and also throw an exception if a mod attempts to enable a flag after dynamic registries are frozen.

rinOfTheStars commented:
Going to go through and implement some of the changes suggested by @Pyrofab in a moment. Probably going to have both varargs for the flags _and_ the current system, in order to have similarities to both QSL's static registry flags and FAPI's dynamic registry flags. 

Disabling flags will need to check if the dynamic registries are frozen, and if they are, refuse to change the state of the flag on that registry. This should stop desyncs, hopefully.

rinOfTheStars committed:
Implemented some changes suggested by Pyrofab

- Added protected helper method DynamicMetaRegistryImpl#isFrozen for use in DynamicRegistryFlagManager
- Added DynamicMetaRegistryImpl#LOGGER with the name "quilt_dynamic_registries" as a logger for catching exceptions related to the setting of flags
- Removed the ability to disable dynamic registry flags, changed the signature of the related flag-enabling methods from enableFlag -> setFlag (on both DynamicRegistryFlag and DynamicRegistryFlagManager)
- DynamicRegistryFlagManager#setFlag now throws an IllegalStateException if a mod attempts to enable a flag when dynamic registries are frozen. This is caught and printed to the logger mentioned prior in DynamicRegistryFlag calls
- Added varargs for setting DynamicRegistryFlag values on dynamic registry creation
- Improved DynamicRegistryFlag#OPTIONAL javadocs slightly

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/dynamic/DynamicRegistryFlag.java:
```diff
@@ -0,0 +1,75 @@
+
+/**
+ * Flags that can be set on dynamic registries to define their behavior.
+ * <p>All flags default to being disabled/turned off, and can be enabled or disabled using static methods in this class.
```
"can be enabled or disabled"? it could be "can be toggled"! :p
it does help reducing the redundancy there;

```suggestion
 * <p>
 * All flags default to being turned off, and can be toggled using static methods in this class.
```

rinOfTheStars replied:
The wording here is actually wrong, since they can only be enabled now. Going to fix it in a commit after the upcoming one that adds most of your other style suggestions

rinOfTheStars replied:
Should be good now

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/dynamic/DynamicRegistryFlag.java:
```diff
@@ -0,0 +1,75 @@
+	 * to (logical) clients for synchronization in multiplayer contexts.
+	 * <p>
+	 * <b>Note:</b> This flag is intended only for synchronized dynamic registries. On non-synced dynamic registries, this flag does nothing.
+	 * <p></p>
```
```suggestion
	 * <p>
```

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/dynamic/DynamicRegistryFlag.java:
```diff
@@ -0,0 +1,75 @@
+	OPTIONAL;
+
+	/**
+	 * Enables a specific flag on a dynamic registry.
+	 * @param registryId the value id ({@link RegistryKey#getValue()}) of the target dynamic registry
```
```suggestion
	 * Enables a specific flag on a dynamic registry.
	 *
	 * @param registryId the value id ({@link RegistryKey#getValue()}) of the target dynamic registry
```

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/dynamic/DynamicRegistryFlag.java:
```diff
@@ -0,0 +1,75 @@
+	}
+
+	/**
+	 * Checks if a dynamic registry has the {@link DynamicRegistryFlag#OPTIONAL} flag enabled on it.
+	 * @param registryId the value id ({@link RegistryKey#getValue()}) of the dynamic registry to check
```
```suggestion
	 * Checks if a dynamic registry has the {@link DynamicRegistryFlag#OPTIONAL} flag enabled on it.
	 *
	 * @param registryId the value id ({@link RegistryKey#getValue()}) of the dynamic registry to check
```

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/dynamic/DynamicMetaRegistryImpl.java:
```diff
@@ -55,4 +63,8 @@ public static <E> void registerSynced(RegistryKey<? extends Registry<E>> ref, Co
 	}
+
+	static boolean isFrozen() {
+		return frozen;
```
just a comment: hnnghhngh i hate the lack of a `this.` there but i can't fight fateeeeeeeee, it's a static field ;-;

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/dynamic/DynamicRegistryFlagManager.java:
```diff
@@ -0,0 +1,43 @@
+			MultimapBuilder.hashKeys().enumSetValues(DynamicRegistryFlag.class).build();
+
+	public static void setFlag(Identifier registryId, DynamicRegistryFlag flag) throws IllegalStateException {
+		if (DynamicMetaRegistryImpl.isFrozen()) throw new IllegalStateException("Dynamic registries are frozen, and thus flags cannot be changed!");
+		if (!DynamicMetaRegistryImpl.isModdedRegistryId(registryId)) return;
+		DYNAMIC_REGISTRY_FLAGS.put(registryId, flag);
```
While there are people against inlined `return`s, I'm personally not, as long as the line is of a reasonable size;
However, at the same time? I believe that `throws` should be bracketed; with the last nitpick being adding a newline between the last if and the put;
So we have:

```suggestion
		if (DynamicMetaRegistryImpl.isFrozen()) {
			throw new IllegalStateException("Dynamic registries are frozen, and thus flags cannot be changed!");
		}

		if (!DynamicMetaRegistryImpl.isModdedRegistryId(registryId)) return;

		DYNAMIC_REGISTRY_FLAGS.put(registryId, flag);
```

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/dynamic/DynamicRegistryFlagManager.java:
```diff
@@ -0,0 +1,43 @@
+	public static boolean isOptional(Identifier registryId) {
+		if (DYNAMIC_REGISTRY_FLAGS.containsKey(registryId)) {
+			return DYNAMIC_REGISTRY_FLAGS.get(registryId).contains(DynamicRegistryFlag.OPTIONAL);
+		}
+		return false;
```
Nitpick
```suggestion
		}

		return false;
```

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/mixin/DynamicRegistrySyncMixin.java:
```diff
@@ -1,4 +1,5 @@
 /*
+ * Copyright 2016, 2017, 2018, 2019 FabricMC
```
aw..
(dw, it can be kept :p)

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/mixin/DynamicRegistrySyncMixin.java:
```diff
@@ -17,21 +18,52 @@
+		// OPTIONAL
+		if (DynamicRegistryFlagManager.isOptional(entry.key().getValue())) {
+			return entry.value().size() > 0;
+		}
+		return true; //if no flags apply, always return true
```
hm, am not sure what are the general feelings about inline comments, but maybe it's fine here, with just an extra space :p
```suggestion
		}

		return true; // If no flags apply, always return true
```

EnnuiL reviewed (approved):
While I do have style cleanup suggestions in mind, I believe it would get way too nitpicky to have it block a PR (I'd rather have a post-merge commit than that);
The important thing is: the logic looks sound to me; Ennui's Stamp of Approval (Can Skip FCP Edition)

EnnuiL committed:
Merge branch '1.20' into feature/dynamic-registry-flags

rinOfTheStars committed:
Commit most style changes

One of them needs a more specific wording change, so it's going to be done in another commit since that's easier for us

Co-authored-by: Ennui Langeweile <85590273+EnnuiL@users.noreply.github.com>

rinOfTheStars committed:
Make documentation for DynamicRegistryFlag more accurate

rinOfTheStars committed:
Improve the javadocs on DynamicRegistryFlag AGAIN

accidentally dropped a word

guh

rinOfTheStars commented:
I LOVE FORGETTING A SINGLE WORD AND HAVING TO MAKE ANOTHER COMMIT!!!!!

EnnuiL commented:
> I LOVE FORGETTING A SINGLE WORD AND HAVING TO MAKE ANOTHER COMMIT!!!!!

mood

EnnuiL commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/dynamic/DynamicRegistryFlag.java:
```diff
@@ -0,0 +1,77 @@
+
+/**
+ * Flags that can be set on dynamic registries to define their behavior.
+ * <p>All flags are off by default, and can be enabled using static methods in this class or on dynamic registry creation via varargs.
```
```suggestion
 * <p>
 * All flags are off by default, and can be enabled using static methods in this class or on dynamic registry creation via varargs.
```

rinOfTheStars committed:
BUT WAIT, THERE'S MORE!

Hopefully the final style change moment

Co-authored-by: Ennui Langeweile <85590273+EnnuiL@users.noreply.github.com>

EnnuiL committed:
Merge branch '1.20' into feature/dynamic-registry-flags

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/dynamic/DynamicRegistryFlagManager.java:
```diff
@@ -0,0 +1,48 @@
+		if (!DynamicMetaRegistryImpl.isModdedRegistryId(registryId)) return;
+
+		DYNAMIC_REGISTRY_FLAGS.put(registryId, flag);
+	}
```
```suggestion
	}

```

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/dynamic/DynamicRegistryFlagManager.java:
```diff
@@ -0,0 +1,48 @@
+
+@ApiStatus.Internal
+public final class DynamicRegistryFlagManager {
+	private static final Multimap<Identifier, DynamicRegistryFlag> DYNAMIC_REGISTRY_FLAGS =
```
Why is this a multimap?

rinOfTheStars replied:
Future-proofing. Right now, there's only one flag, but if other flags are added nothing needs to be changed here to store them in memory.

rinOfTheStars committed:
Implement style change in DynamicRegistryFlagManager

Co-authored-by: LambdAurora <aurora42lambda@gmail.com>

Pyrofab commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/dynamic/DynamicMetaRegistry.java:
```diff
@@ -48,8 +48,8 @@ public final class DynamicMetaRegistry {
 	 */
-	public static <E> void register(RegistryKey<? extends Registry<E>> key, Codec<E> entryCodec) {
-		DynamicMetaRegistryImpl.register(key, entryCodec);
+	public static <E> void register(RegistryKey<? extends Registry<E>> key, Codec<E> entryCodec, DynamicRegistryFlag... flags) {
```
This change is not binary compatible, we may want to add an overload for the flagless version

rinOfTheStars replied:
Fixed as of latest commit. Thanks for the catch, didn't know varargs would cause binary incompatibility like that!

rinOfTheStars committed:
Fix binary incompatibility hopefully?
- Added back methods to DynamicMetaRegistry with the same signature as the old ones to avoid binary incompatibility. These are marked as deprecated and should be removed in a breaking/major QSL release.

rinOfTheStars committed:
Merge branch '1.20' into feature/dynamic-registry-flags

rinOfTheStars committed:
Remove deprecation annotations on compat. method overloads in DynamicMetaRegistry

rinOfTheStars committed:
Merge remote-tracking branch 'origin/feature/dynamic-registry-flags' into feature/dynamic-registry-flags

EnnuiL committed:
Merge branch '1.20' into feature/dynamic-registry-flags
